driver/rope_driver.hpp

```diff
@@ -0,0 +1,488 @@
+
+        yhost[o] = val;
+    }
+
+    return ret;
+}
+
+template <typename Tgpu, typename Tcheck>
+int32_t mloRoPEBackwardRunHost(miopenTensorDescriptor_t dyDesc,
+                               miopenTensorDescriptor_t cosDesc,
+                               miopenTensorDescriptor_t sinDesc,
+                               miopenTensorDescriptor_t dxDesc,
+                               Tgpu* dy,
+                               Tgpu* cos,
+                               Tgpu* sin,
+                               Tcheck* dxhost)
+{
+    auto dy_dims = miopen::deref(dyDesc).GetLengths();
+    auto input_numel =
+        std::accumulate(dy_dims.begin(), dy_dims.end(), 1LL, std::multiplies<int64_t>());
+    auto cos_dims = miopen::deref(cosDesc).GetLengths();
+    auto rotary_numel =
+        std::accumulate(cos_dims.begin(), cos_dims.end(), 1LL, std::multiplies<int64_t>());
+
+    int32_t ret = 0;
+
+    for(size_t o = 0; o < input_numel; o++)
+    {
+        size_t freq_idx = o % rotary_numel;
+
+        Tcheck output_grad = static_cast<Tcheck>(dy[o]);
+        Tcheck output_grad_rotate_half =
+            (o % 2 == 0) ? static_cast<Tcheck>(dy[o + 1]) : static_cast<Tcheck>(-dy[o - 1]);
+
+        Tcheck cos_val = static_cast<Tcheck>(cos[freq_idx]);
+        Tcheck sin_val = (freq_idx % 2 == 0) ? static_cast<Tcheck>(sin[freq_idx + 1])
+                                             : static_cast<Tcheck>(sin[freq_idx - 1]);
+
+        Tcheck val = (output_grad * cos_val) + (output_grad_rotate_half * sin_val);
+
+        dxhost[o] = val;
+    }
+
+    return ret;
+}
+
+template <typename Tgpu, typename Tref>
+class RoPEDriver : public Driver
+{
+public:
+    RoPEDriver() : Driver()
+    {
+        miopenCreateTensorDescriptor(&xDesc);
+        miopenCreateTensorDescriptor(&cosDesc);
+        miopenCreateTensorDescriptor(&sinDesc);
+        miopenCreateTensorDescriptor(&yDesc);
+        miopenCreateTensorDescriptor(&dyDesc);
+        miopenCreateTensorDescriptor(&dxDesc);
+
+        data_type = miopen_type<Tgpu>{};
+    }
+
+    int AddCmdLineArgs() override;
+    int ParseCmdLineArgs(int argc, char* argv[]) override;
+    InputFlags& GetInputFlags() override { return inflags; }
+
+    int GetandSetData() override;
+    std::vector<int> GetInputTensorLengthsFromCmdLine();
+
+    int AllocateBuffersAndCopy() override;
+
+    int RunForwardGPU() override;
+    int RunForwardCPU();
+
+    int RunBackwardGPU() override;
+    int RunBackwardCPU();
+
+    Tref GetTolerance();
+    int VerifyBackward() override;
+    int VerifyForward() override;
+    ~RoPEDriver() override
+    {
+
+        miopenDestroyTensorDescriptor(xDesc);
+        miopenDestroyTensorDescriptor(cosDesc);
+        miopenDestroyTensorDescriptor(sinDesc);
+        miopenDestroyTensorDescriptor(yDesc);
+        miopenDestroyTensorDescriptor(dyDesc);
+        miopenDestroyTensorDescriptor(dxDesc);
+    }
+
+private:
+    InputFlags inflags;
+
+    int forw;
```

Comment:
Seems it is unused.

Reply:
It is used here.
https://github.com/ROCm/MIOpen/blob/d208a9a3351eb9048a4a2b3f9d34fefbae0f6fda/driver/main.cpp#L84

Reply:
But who initializes it? Moreover it's a string and I don't recall that MIOpen (and C++) can extract class members by the string name in the runtime.
Also `drv->GetInputFlags()` returns `inflags`, not `RoPEDriver`.
It is still unused.

Reply:
You're right. I misunderstood.